src/info/git/mod.rs

```diff
@@ -0,0 +1,338 @@
+use self::metrics::GitMetrics;
+use self::sig::Sig;
+use crate::cli::{CliOptions, MyRegex};
+use anyhow::Result;
+use gix::bstr::ByteSlice;
+use gix::bstr::{BString, Utf8Error};
+use gix::object::tree::diff::change::Event;
+use gix::object::tree::diff::Action;
+use gix::objs::tree::EntryMode;
+use gix::prelude::ObjectIdExt;
+use gix::traverse::commit::Sorting;
+use gix::{Commit, ObjectId};
+use regex::Regex;
+use std::collections::HashMap;
+use std::str::FromStr;
+use std::sync::atomic::{AtomicBool, AtomicUsize, Ordering};
+use std::sync::mpsc::{channel, Sender};
+use std::sync::Arc;
+use std::thread::JoinHandle;
+
+pub mod metrics;
+mod sig;
+
+pub fn traverse_commit_graph(repo: &gix::Repository, options: &CliOptions) -> Result<GitMetrics> {
+    let mut time_of_most_recent_commit = None;
+    let mut time_of_first_commit = None;
+    let mut number_of_commits_by_signature: HashMap<Sig, usize> = HashMap::new();
+    let mailmap = repo.open_mailmap();
+    let bot_regex_pattern = get_no_bots_regex(&options.info.no_bots)?;
+    let has_commit_graph_traversal_ended = Arc::new(AtomicBool::default());
+    let total_number_of_commits = Arc::new(AtomicUsize::default());
+
+    let commit_iter = repo
+        .head_commit()?
+        .id()
+        .ancestors()
+        .sorting(Sorting::ByCommitTimeNewestFirst)
+        .all()?;
+
+    let (churn_thread, churn_tx) = get_churn_channel(
+        repo,
+        &has_commit_graph_traversal_ended,
+        &total_number_of_commits,
+        options.info.churn_pool_size,
+    )?;
+
+    let num_threads = std::thread::available_parallelism()
```

Comment:
And all that is to say, if you feel differently about thread allocation, feel free to change it, of course. The logic could probably be better anyway as I test on M1 Pro and on this machine it's fine, but a quad-core machine might feel differently about the over-commit on threads and I didn't test on these.

Reply:
Very well, this is well thought out. I was wondering why you chose to use multiple threads for author computation, but this does make sense. The main thread has to wait for it to finish, and with the commit-graph, the traversal is very fast, so we don't want to lose what we gained with the commit-graph; the churn computation is on a best-effort basis, so it won't affect the speed.

Reply:
One last question/confirmation, @Byron: It appears that the performance improvement achieved through parallelization is only possible when a commit graph is available, and when we limit ourselves to some commit information such as author and time. Is my understanding correct, and if so, why is that the case? Does fetching commit information from storage instead of the commit-graph involve some form of locking that negates any potential benefits of using multiple threads in the loop?

Reply:
Yes, this will only kick in if a commit-graph is present. Without it, commit traversals are naturally bound to be done on a single thread - to get the parent commits, the commit object needs to be decoded (as a blob/`&[u8]), and that's very time consuming. Right now `gitoxide` can do about 130k commits per second, but that's quite slow in a repo with a million commits. With a commit-graph, we are not bound by this restriction anymore.

> Does fetching commit information from storage instead of the commit-graph involve some form of locking that negates any potential benefits of using multiple threads in the loop?

No, there is no locking and accessing objects from object storage scales perfectly with cores (there is no lock), but as mentioned before there isn't anything one can do in terms of parallelization with single-trunk histories. In theory it would be possible to parallelize across branches but it's messy, the threads have to synchronize and you know what that means. With two or three threads, maybe there is some gain to be had though.

I hope this works as an answer :).

Reply:
Very clear, but what I did not understand was why `no commit graph + multithreading (for author info) ~= no performance gain`, but this is due to these [two lines](https://github.com/o2sh/onefetch/blob/75dff28e7b8d8ef46b348a2bd099f80f6cae8073/src/info/git/mod.rs#LL145C1-L146C47).

If we don't `repo.object_cache_size(0);` - in the case where there is no commit graph - we can actually gain a lot of performance because the commit object would have already been cached previously - based on what you said about the need to decode the object to get the parent commits.

As a result, maybe we should multithread even in the nominal case - no commit graph - but no override the repo's object each ? 

What do you think, @Byron?

Reply:
Very clear, but I was not suggesting multithreading the commit's history traversal. My question relates to the current use of multithreading in the code, which is employed to retrieve the commit's author information. Specifically, if I manually set can_use_author_threads to true and use_commit_graph(false), I don't observe any performance improvement, and when running time onefetch, I notice a high "system" value. I attempted removing repo.object_cache_size(0); in [this line](https://github.com/o2sh/onefetch/blob/75dff28e7b8d8ef46b348a2bd099f80f6cae8073/src/info/git/mod.rs#L146).
without any effect.

Reply:
Very clear, but I was not suggesting multithreading the commit traversal. My question relates to the current use of multithreading in the code, which is employed to retrieve the commit's author information in parallel. For example, if I manually set `can_use_author_threads` to true and `use_commit_graph(false)`, I don't see any performance improvement compared to using a single thread, and when running `time onefetch`, I notice a high "system" value which seems to go against:

>accessing objects from object storage scales perfectly with cores

I attempted removing `repo.object_cache_size(0);` in [this line](https://github.com/o2sh/onefetch/blob/75dff28e7b8d8ef46b348a2bd099f80f6cae8073/src/info/git/mod.rs#L146) without any effect.

Reply:
Yeah, you can scrap my previous message. 😅 Upon re-reading it now, I must have been tired. Sorry about that, and thanks again for the PR! 🙏😊